bgpd/bgp_aspath.c

```diff
@@ -2096,3 +2096,146 @@ void aspath_print_all_vty(struct vty *vty)
 				       void *))aspath_show_all_iterator,
 		     vty);
 }
+
+static unsigned int bgp_aggr_aspath_hash_key(void *data)
+{
+	struct aspath *aspath = data;
+
+	return jhash(aspath->str, aspath->str_len, 0xdcba4321);
+}
+
+static bool bgp_aggr_aspath_hash_cmp(const void *arg1, const void *arg2)
+{
+	const struct assegment *seg1 = ((const struct aspath *)arg1)->segments;
+	const struct assegment *seg2 = ((const struct aspath *)arg2)->segments;
+
+	while (seg1 || seg2) {
+		int i;
+
+		if ((!seg1 && seg2) || (seg1 && !seg2))
+			return false;
+		if (seg1->type != seg2->type)
+			return false;
+		if (seg1->length != seg2->length)
+			return false;
+		for (i = 0; i < seg1->length; i++)
+			if (seg1->as[i] != seg2->as[i])
+				return false;
+		seg1 = seg1->next;
+		seg2 = seg2->next;
+	}
+	return true;
+}
+
+static struct aspath *bgp_aggr_aspath_lookup(struct bgp_aggregate *aggregate,
+					     struct aspath *aspath)
+{
+	return hash_lookup(aggregate->aspath_hash, aspath);
+}
+
+static void *bgp_aggr_aspath_hash_alloc(void *p)
+{
+	struct aspath *ref = (struct aspath *)p;
+	struct aspath *aspath = NULL;
+
+	aspath = aspath_dup(ref);
+	return aspath;
+}
+
+static void bgp_aggr_aspath_prepare(struct hash_backet *hb, void *arg)
+{
+	struct aspath *asmerge = NULL;
+	struct aspath *hb_aspath = hb->data;
+	struct aspath **aggr_aspath = arg;
+
+	if (*aggr_aspath) {
+		asmerge = aspath_aggregate(*aggr_aspath, hb_aspath);
+		aspath_free(*aggr_aspath);
+		*aggr_aspath = asmerge;
+	} else
+		*aggr_aspath = aspath_dup(hb_aspath);
+}
+
+void bgp_aggr_aspath_remove(void *arg)
+{
+	struct aspath *aspath = arg;
+
+	aspath_free(aspath);
+}
+
+void bgp_compute_aggregate_aspath(struct bgp_aggregate *aggregate,
+				  struct aspath *aspath)
+{
+	struct aspath *aggr_aspath = NULL;
+
+	if ((aggregate == NULL) || (aspath == NULL))
+		return;
+
+	/* Create hash if not already created.
+	 */
+	if (aggregate->aspath_hash == NULL) {
+		aggregate->aspath_hash = hash_create(
+					bgp_aggr_aspath_hash_key,
+					bgp_aggr_aspath_hash_cmp,
+					"BGP Aggregator as-path hash");
+		assert(aggregate->aspath_hash != NULL);
```

Comment:
hash_create doesn't fail or the entire program crashes no need for an assert

Reply:
I've removed the assert.